metaflow/plugins/deployer.py

```diff
@@ -0,0 +1,206 @@
+
+
+def get_lower_level_group(
+    api, top_level_kwargs: Dict, _type: str, deployer_kwargs: Dict
+):
+    if _type is None:
+        raise ValueError(
+            "ConcreteDeployer doesn't have a 'TYPE' to target. Please use a sub-class of ConcreteDeployer."
+        )
+    return getattr(api(**top_level_kwargs), _type)(**deployer_kwargs)
+
+
+class Deployer(object):
+    methods = {}
+
+    def __init__(
+        self,
+        flow_file: str,
+        show_output: bool = False,
+        profile: Optional[str] = None,
+        env: Optional[Dict] = None,
+        cwd: Optional[str] = None,
+        **kwargs
+    ):
+        self.flow_file = flow_file
+        self.show_output = show_output
+        self.profile = profile
+        self.env = env
+        self.cwd = cwd
+        self.top_level_kwargs = kwargs
+
+        from metaflow.plugins import CONCRETE_DEPLOYER_PROVIDERS
+
+        for provider_class in CONCRETE_DEPLOYER_PROVIDERS:
+            method_name = provider_class.TYPE.replace("-", "_")
+            Deployer.methods[method_name] = provider_class
+            setattr(Deployer, method_name, self.make_function(provider_class))
+
+    def make_function(self, deployer_class):
+        def f(self, **deployer_kwargs):
+            return deployer_class(
+                deployer_kwargs=deployer_kwargs,
+                flow_file=self.flow_file,
+                show_output=self.show_output,
+                profile=self.profile,
+                env=self.env,
+                cwd=self.cwd,
+                **self.top_level_kwargs
+            )
+
+        return f
+
+
+class TriggeredRun(object):
+    def __init__(
+        self,
+        deployer: "ConcreteDeployer",
+        content: str,
+    ):
+        self.deployer = deployer
+        content_json = json.loads(content)
+        self.metadata_for_flow = content_json.get("metadata")
+        self.pathspec = content_json.get("pathspec")
+        self.name = content_json.get("name")
+
+    def __contains__(self, key: str):
+        return getattr(self, key, None) is not None
+
+    def get(self, key: str, default=None) -> Optional[Any]:
+        return getattr(self, key, default)
+
+    def _update_env(self, env):
+        for k, v in env.items():
+            if isinstance(v, property):
+                setattr(self.__class__, k, v)
+            elif callable(v):
+                setattr(self, k, v.__get__(self))
+            else:
+                setattr(self.__class__, k, property(fget=lambda _, v=v: v))
+
+    @property
+    def run(self):
+        from metaflow import Run, metadata
+
+        clear_and_set_os_environ(self.deployer.old_env)
+        metadata(self.metadata_for_flow)
+        try:
+            return Run(self.pathspec, _namespace_check=False)
+        except MetaflowNotFound:
+            return None
+
+
+class DeployedFlow(object):
+    def __init__(self, deployer: "ConcreteDeployer"):
+        self.deployer = deployer
+
+    def __contains__(self, key: str):
```

Comment:
what is the use of `__contains__` and `get` here?

Reply:
this was inherited from the implementation of the `current` object, removing...